src/CoreWCF.Primitives/tests/Extensibility/CallContextInitializersTests.cs

```diff
@@ -0,0 +1,113 @@
+            factory.Close();
+            TestHelper.CloseServiceModelObjects((System.ServiceModel.Channels.IChannel)channel, factory);
+        }
+
+        [Fact]
+        public static void CallContextInitializersCalledMulitple()
+        {
+            const int timesToCall = 10;
+            var callContextInitializer = new TestCallContextInitializers();
+            var endpointBehavior = new TestEndpointBehavior(callContextInitializer);
+
+            System.ServiceModel.ChannelFactory<ISimpleService> factory = ExtensibilityHelper.CreateChannelFactory<SimpleService, ISimpleService>((CoreWCF.ServiceHostBase serviceHostBase) =>
+            {
+                foreach (var endpoint in serviceHostBase.Description.Endpoints)
+                {
+                    endpoint.EndpointBehaviors.Add(endpointBehavior);
+                }
+            });
+            factory.Open();
+            ISimpleService channel = factory.CreateChannel();
+            foreach (int dummy in Enumerable.Range(0, timesToCall))
+            {
+                string echo = channel.Echo("hello");
+                Assert.Equal("hello", echo);
+            }
+
+            Assert.Equal(timesToCall, callContextInitializer.BeforeInvokeCount);
+            Assert.Equal(timesToCall, callContextInitializer.AfterInvokeCount);
+
+            ((System.ServiceModel.Channels.IChannel)channel).Close();
+            factory.Close();
+            TestHelper.CloseServiceModelObjects((System.ServiceModel.Channels.IChannel)channel, factory);
+        }
+    }
+
+    internal class TestEndpointBehavior : IEndpointBehavior
+    {
+        private readonly ICallContextInitializer _callContextInitializer;
+
+        public TestEndpointBehavior(ICallContextInitializer callContextInitializer)
+        {
+            _callContextInitializer = callContextInitializer;
+        }
+
+        public void AddBindingParameters(ServiceEndpoint endpoint, BindingParameterCollection bindingParameters) { }
+
+        public void ApplyClientBehavior(ServiceEndpoint endpoint, ClientRuntime clientRuntime) { }
+
+        public void ApplyDispatchBehavior(ServiceEndpoint endpoint, EndpointDispatcher endpointDispatcher)
+        {
+            foreach (DispatchOperation op in endpointDispatcher.DispatchRuntime.Operations)
+            {
+                op.CallContextInitializers.Add(_callContextInitializer);
+            }
+        }
+
+        public void Validate(ServiceEndpoint endpoint) { }
+    }
+
+    internal class TestCallContextInitializers : ICallContextInitializer
+    {
+        public int AfterInvokeCount;
+        public int BeforeInvokeCount;
+
+        public void AfterInvoke(object correlationState)
+        {
+            AfterInvokeCount++;
+        }
+
+        public object BeforeInvoke(InstanceContext instanceContext, IClientChannel channel, Message message)
+        {
+            BeforeInvokeCount++;
+            return new object();
+        }
```

Comment:
Can you add a test to ensure that the returned correlation object is passed into AfterInvoke?